text/0000-delegation.md

```diff
@@ -0,0 +1,700 @@
+[object_composition]: https://en.wikipedia.org/wiki/Composition_over_inheritance
+[initial_rfc]:
+https://github.com/rust-lang/rfcs/pull/1406
+[design_discussion]:
+https://internals.rust-lang.org/t/3-weeks-to-delegation-please-help/5742
+[internals_thread]: https://internals.rust-lang.org/t/new-rfc-for-delegation-anyone-interested-in-contributing/6644
+
+# Guide-level explanation
+[guide]: #guide
+
+In Rust, we prefer composition over inheritence for code reuse. For common cases, we make this convenient with delegation syntax sugar.
+
+Whenever you have a struct `S` with a member field `f` of type `F` and `F` already implements a trait `TR`, you can delegate the implementation of `TR` for `S` to `f` using the keyword `delegate`:
+
+```rust
+impl TR for S {
+    delegate * to f;
+}
+```
+
+This is pure sugar, and does exactly the same thing as if you “manually delegated” all the items of `TR` like this:
+
+```rust
+impl TR for S {
+    type Item = <F as TR>::Item;
+    // ...
+    const MAX = <F as TR>::MAX;
+    // ...
+    fn foo(&self) -> u32 {
+        self.f.foo()
+    }
+    fn bar(&self, x: u32, y: u32, z: u32) -> u32 {
+        self.f.bar(x, y, z)
+    }
+    // ...
+}
+```
+
+To delegate most of a trait, rather than all of it, simply `delegate *` and then write the manual implementations for the items you don’t want to delegate.
+
+```rust
+impl TR for S {
+    delegate * to f;
+
+    fn foo(&self) -> u32 {
+        42
+    }
+}
+```
+
+Aside from the implementation of `foo()`, this has exactly the same meaning as the first example.
+
+If you only want to delegate specific items, rather than “all” or “most” items, then replace `*` with a comma-separated list of only the items you want to delegate. Since it’s possible for types and functions to have the same name, the items must be prefixed with `fn`, `const` and `type` as appropriate.
+
+```rust
+impl TR for S {
+    delegate fn foo, fn bar, const MAX, type Item
+        to f;
+}
+```
+
+This also has the exact same meaning as the first example. While `to field_name;` is not required to be on a new line, this is the `rustfmt` default when delegating individual trait items, as it makes it easy to visually identify the field being delegated to.
+
+If you want to delegate some methods to one field and some to another, simply write multiple `delegate` statements:
+
+```rust
+impl TR for S {
+    delegate * to field_one;
+    delegate fn foo, const MAX, type Item
+        to field_two;
+}
+```
+
+This delegates `foo`, `MAX`, and `Item` to `field_two`. Everything else is delegated to `field_one`.
+
+In addition to "regular" structs with named fields, you can delegate to indexed fields on a tuple struct. This first field name is `0`, which we're used to writing as `self.0`:
+
+```rust
+struct AddOnlyCounter(u32);
+
+impl AddOnlyCounter {
+    fn new() -> Self {
+        Self { 0 }
+    }
+}
+
+impl AddAssign for AddOnlyCounter {
+    delegate * to 0;
+}
```

Comment:
I was already thinking of suggesting that it be `delegate <list> to self.<field>`, and this solidified the feeling.  (`delegate fn hash to 0` looks like it'd do `0.hash()` to me.) I like there being restrictions as a starting point, but I wouldn't be surprised at all for this to get expanded, and expanding to expressions (perhaps a subset thereof, but note that `unimplemented!()` would just work if it took anything) seems like the most obvious one, so I think the syntax should at least look like a normal expression from the beginning.  (Readers: feel free to :+1:(:-1:) if you would(n't) like me to make a PR for that.)